src/System.Private.ServiceModel/tests/Scenarios/Extensibility/WebSockets/WebSocketTests.cs

```diff
@@ -234,4 +234,230 @@ public static void WebSocketHttpsDuplexBinaryStreamed()
+
+            // *** CLEANUP *** \\
+            ((ICommunicationObject)client).Close();
+            channelFactory.Close();
+        }
+        finally
+        {
+            // *** ENSURE CLEANUP *** \\
+            ScenarioTestHelpers.CloseCommunicationObjects((ICommunicationObject)client, channelFactory);
+            clientReceiver.Dispose();
+        }
+    }
+
+    [Fact]
+    [OuterLoop]
+    [ActiveIssue(468)]
+    public static void WebSocketHttpDuplexTextStreamed()
+    {
+        NetHttpBinding binding = null;
+        ClientReceiver clientReceiver = null;
+        InstanceContext context = null;
+        DuplexChannelFactory<IWSDuplexService> channelFactory = null;
+        IWSDuplexService client = null;
+        FlowControlledStream uploadStream = null;
+
+        try
+        {
+            // *** SETUP *** \\
+            binding = new NetHttpBinding();
+            binding.WebSocketSettings.TransportUsage = WebSocketTransportUsage.Always;
+            binding.MaxReceivedMessageSize = ScenarioTestHelpers.DefaultMaxReceivedMessageSize;
+            binding.MaxBufferSize = ScenarioTestHelpers.DefaultMaxReceivedMessageSize;
+            binding.TransferMode = TransferMode.Streamed;
+            binding.MessageEncoding = NetHttpMessageEncoding.Text;
+
+            clientReceiver = new ClientReceiver();
+            context = new InstanceContext(clientReceiver);
+
+            channelFactory = new DuplexChannelFactory<IWSDuplexService>(context, binding, Endpoints.WebSocketHttpDuplexTextStreamed_Address);
+            client = channelFactory.CreateChannel();
+
+            // *** EXECUTE *** \\
+            using (Stream stream = client.DownloadStream())
+            {
+                int readResult;
+                // Read from the stream, 1000 bytes at a time.
+                byte[] buffer = new byte[1000];
+                do
+                {
+                    readResult = stream.Read(buffer, 0, buffer.Length);
+                }
+                while (readResult != 0);
+            }
+
+            uploadStream = new FlowControlledStream();
+            uploadStream.ReadThrottle = TimeSpan.FromMilliseconds(500);
+            uploadStream.StreamDuration = TimeSpan.FromSeconds(1);
+
+            client.UploadStream(uploadStream);
+            client.StartPushingStream();
+            // Wait for the callback to get invoked before telling the service to stop streaming.
+            // This ensures we can read from the stream on the callback while the NCL layer at the service
+            // is still writing the bytes from the stream to the wire.  
+            // This will deadlock if the transfer mode is buffered because the callback will wait for the
+            // stream, and the NCL layer will continue to buffer the stream until it reaches the end.
+
+            clientReceiver.ReceiveStreamInvoked.WaitOne();
+            clientReceiver.ReceiveStreamInvoked.Reset();
```

Comment:
All WaitOne's should specify a max timeout, or we risk hanging here forever -- triply painful on Linux.  It should also test the return value of WaitOne to fail if it did timeout.  I'd use MaxTestTimeout here too.